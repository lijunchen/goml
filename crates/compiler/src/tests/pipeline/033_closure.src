fn test() {
    let y = 3 in
    let z = 5 in
    let f = |x: int| x * y * z in
    let _ = string_println(int_to_string(f(2))) in
    string_println(int_to_string(f(3)))
}

fn call_int_id(f : (int) -> int, v: int) -> int {
    f(v)
}

enum IntList {
    Nil,
    Cons(int, IntList),
}

struct Point {
    x: int,
    y: int,
}

fn main() {
    let base = 5 in
    let add_base = |x: int| x + base in
    let result = add_base(7) in
    let printer = |prefix: string, value: int| {
        let message = string_add(prefix, int_to_string(value)) in
        string_println(message)
    } in
    let _ = printer("result: ", result) in
    let unused = |y: int| y + result in
    let no_capture = |z: int| z * 2 in
    let doubled = no_capture(3) in
    let _ = string_println(int_to_string(doubled)) in
    let _ = test() in

    let list123 = Cons(1, Cons(2, Cons(3, Nil))) in

    let point = Point { x: 10, y: 20 } in

    let play_list_and_point = || {
        match list123 {
            Nil => string_println("Empty list"),
            Cons(head, tail) => {
                let _ = string_println(int_to_string(head)) in
                match point {
                    Point { x: x, y: y } => {
                        let _ = string_println(string_add("Point: (", string_add(int_to_string(x), string_add(", ", string_add(int_to_string(y), ")"))))) in
                        ()
                    }
                }
            }
        }
    } in

    let _ = play_list_and_point() in
    ()
}
