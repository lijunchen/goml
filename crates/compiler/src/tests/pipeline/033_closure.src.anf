fn test() -> unit {
  let y/0 = 3 in
  let z/1 = 5 in
  let f/3 = closure_env_f_0 { y_0: y/0, z_1: z/1 } in
  let t24 = __closure_apply11(f/3, 2) in
  let t23 = int_to_string(t24) in
  let mtmp0 = string_println(t23) in
  let t26 = __closure_apply11(f/3, 3) in
  let t25 = int_to_string(t26) in
  string_println(t25)
}

fn call_int_id(f/4: (int) -> int, v/5: int) -> int {
  f/4(v/5)
}

fn main() -> unit {
  let base/6 = 5 in
  let add_base/8 = closure_env_add_base_1 { base_0: base/6 } in
  let result/9 = __closure_apply13(add_base/8, 7) in
  let printer/13 = closure_env_printer_2 {} in
  let mtmp1 = __closure_apply15(printer/13, "result: ", result/9) in
  let unused/15 = closure_env_unused_3 { result_0: result/9 } in
  let no_capture/17 = closure_env_no_capture_4 {} in
  let doubled/18 = __closure_apply19(no_capture/17, 3) in
  let t27 = int_to_string(doubled/18) in
  let mtmp2 = string_println(t27) in
  let mtmp3 = test() in
  let t30 = Tag_0 in
  let t29 = IntList::Cons(3, t30) in
  let t28 = IntList::Cons(2, t29) in
  let list123/19 = IntList::Cons(1, t28) in
  let point/20 = Point { x: 10, y: 20 } in
  let play_list_and_point/25 = closure_env_play_list_and_point_5 { list123_0: list123/19, point_1: point/20 } in
  let mtmp10 = __closure_apply21(play_list_and_point/25) in
  ()
}

fn __closure_apply11(env12: closure_env_f_0, x/2: int) -> int {
  let y/0 = closure_env_f_0.y_0(env12) in
  let z/1 = closure_env_f_0.z_1(env12) in
  let t31 = int_mul(x/2, y/0) in
  int_mul(t31, z/1)
}

fn __closure_apply13(env14: closure_env_add_base_1, x/7: int) -> int {
  let base/6 = closure_env_add_base_1.base_0(env14) in
  int_add(x/7, base/6)
}

fn __closure_apply15(env16: closure_env_printer_2, prefix/10: string, value/11: int) -> unit {
  let t32 = int_to_string(value/11) in
  let message/12 = string_add(prefix/10, t32) in
  string_println(message/12)
}

fn __closure_apply17(env18: closure_env_unused_3, y/14: int) -> int {
  let result/9 = closure_env_unused_3.result_0(env18) in
  int_add(y/14, result/9)
}

fn __closure_apply19(env20: closure_env_no_capture_4, z/16: int) -> int {
  int_mul(z/16, 2)
}

fn __closure_apply21(env22: closure_env_play_list_and_point_5) -> unit {
  let list123/19 = closure_env_play_list_and_point_5.list123_0(env22) in
  let point/20 = closure_env_play_list_and_point_5.point_1(env22) in
  match list123/19 {
    Tag_0 => {
      string_println("Empty list")
    },
    Tag_1 => {
      let x4 = IntList::Cons._0(list123/19) in
      let x5 = IntList::Cons._1(list123/19) in
      let tail/22 = x5 in
      let head/21 = x4 in
      let t33 = int_to_string(head/21) in
      let mtmp6 = string_println(t33) in
      let x7 = Point.x(point/20) in
      let x8 = Point.y(point/20) in
      let y/24 = x8 in
      let x/23 = x7 in
      let t36 = int_to_string(x/23) in
      let t39 = int_to_string(y/24) in
      let t38 = string_add(t39, ")") in
      let t37 = string_add(", ", t38) in
      let t35 = string_add(t36, t37) in
      let t34 = string_add("Point: (", t35) in
      let mtmp9 = string_println(t34) in
      ()
    },
  }
}