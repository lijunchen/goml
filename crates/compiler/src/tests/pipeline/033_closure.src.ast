fn main() {
    let base = 5 in
    let add_base = |x: int| x + base in
    let result = add_base(7) in
    let printer = |prefix: string, value: int| let message = string_add(prefix, int_to_string(value)) in
    string_println(message) in
    let _ = printer("result: ", result) in
    let unused = |y: int| y + result in
    let no_capture = |z: int| z * 2 in
    let doubled = no_capture(3) in
    let _ = string_println(int_to_string(doubled)) in
    ()
}

