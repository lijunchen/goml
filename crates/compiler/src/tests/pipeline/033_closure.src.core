fn test() -> unit {
  let y/0 = 3 in
  let z/1 = 5 in
  let f/3 = |x/2: int| => int_mul(int_mul(x/2, y/0), z/1) in
  let mtmp0 = string_println(int_to_string(f/3(2))) in
  string_println(int_to_string(f/3(3)))
}

fn call_int_id(f/4: (int) -> int, v/5: int) -> int {
  f/4(v/5)
}

fn main() -> unit {
  let base/6 = 5 in
  let add_base/8 = |x/7: int| => int_add(x/7, base/6) in
  let result/9 = add_base/8(7) in
  let printer/13 = |prefix/10: string, value/11: int| => let message/12 = string_add(prefix/10, int_to_string(value/11)) in
  string_println(message/12) in
  let mtmp1 = printer/13("result: ", result/9) in
  let unused/15 = |y/14: int| => int_add(y/14, result/9) in
  let no_capture/17 = |z/16: int| => int_mul(z/16, 2) in
  let doubled/18 = no_capture/17(3) in
  let mtmp2 = string_println(int_to_string(doubled/18)) in
  let mtmp3 = test() in
  let list123/19 = IntList::Cons(1, IntList::Cons(2, IntList::Cons(3, IntList::Nil))) in
  let point/20 = Point { x: 10, y: 20 } in
  let play_list_and_point/25 = || => match list123/19 {
    IntList::Nil => {
      string_println("Empty list")
    },
    IntList::Cons(x4, x5) => {
      let x4 = IntList::Cons._0(list123/19) in
      let x5 = IntList::Cons._1(list123/19) in
      let tail/22 = x5 in
      let head/21 = x4 in
      let mtmp6 = string_println(int_to_string(head/21)) in
      let x7 = Point.x(point/20) in
      let x8 = Point.y(point/20) in
      let y/24 = x8 in
      let x/23 = x7 in
      let mtmp9 = string_println(string_add("Point: (", string_add(int_to_string(x/23), string_add(", ", string_add(int_to_string(y/24), ")"))))) in
      ()
    },
  } in
  let mtmp10 = play_list_and_point/25() in
  ()
}