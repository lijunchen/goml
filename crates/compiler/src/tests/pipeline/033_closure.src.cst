FILE@0..738
  FN@0..174
    FnKeyword@0..2 "fn"
    Whitespace@2..3 " "
    Lident@3..7 "test"
    PARAM_LIST@7..10
      LParen@7..8 "("
      RParen@8..9 ")"
      Whitespace@9..10 " "
    BLOCK@10..174
      LBrace@10..11 "{"
      Whitespace@11..16 "\n    "
      EXPR_LET@16..171
        LetKeyword@16..19 "let"
        Whitespace@19..20 " "
        PATTERN_VARIABLE@20..22
          Lident@20..21 "y"
          Whitespace@21..22 " "
        Eq@22..23 "="
        Whitespace@23..24 " "
        EXPR_LET_VALUE@24..26
          EXPR_INT@24..26
            Int@24..25 "3"
            Whitespace@25..26 " "
        InKeyword@26..28 "in"
        Whitespace@28..33 "\n    "
        EXPR_LET_BODY@33..171
          EXPR_LET@33..171
            LetKeyword@33..36 "let"
            Whitespace@36..37 " "
            PATTERN_VARIABLE@37..39
              Lident@37..38 "z"
              Whitespace@38..39 " "
            Eq@39..40 "="
            Whitespace@40..41 " "
            EXPR_LET_VALUE@41..43
              EXPR_INT@41..43
                Int@41..42 "5"
                Whitespace@42..43 " "
            InKeyword@43..45 "in"
            Whitespace@45..50 "\n    "
            EXPR_LET_BODY@50..171
              EXPR_LET@50..171
                LetKeyword@50..53 "let"
                Whitespace@53..54 " "
                PATTERN_VARIABLE@54..56
                  Lident@54..55 "f"
                  Whitespace@55..56 " "
                Eq@56..57 "="
                Whitespace@57..58 " "
                EXPR_LET_VALUE@58..77
                  EXPR_CLOSURE@58..77
                    CLOSURE_PARAM_LIST@58..67
                      Pipe@58..59 "|"
                      CLOSURE_PARAM@59..65
                        PATTERN_VARIABLE@59..60
                          Lident@59..60 "x"
                        Colon@60..61 ":"
                        Whitespace@61..62 " "
                        TYPE_INT@62..65
                          IntKeyword@62..65 "int"
                      Pipe@65..66 "|"
                      Whitespace@66..67 " "
                    EXPR_CLOSURE_BODY@67..77
                      EXPR_BINARY@67..77
                        EXPR_BINARY@67..73
                          EXPR_LIDENT@67..69
                            Lident@67..68 "x"
                            Whitespace@68..69 " "
                          Star@69..70 "*"
                          Whitespace@70..71 " "
                          EXPR_LIDENT@71..73
                            Lident@71..72 "y"
                            Whitespace@72..73 " "
                        Star@73..74 "*"
                        Whitespace@74..75 " "
                        EXPR_LIDENT@75..77
                          Lident@75..76 "z"
                          Whitespace@76..77 " "
                InKeyword@77..79 "in"
                Whitespace@79..84 "\n    "
                EXPR_LET_BODY@84..171
                  EXPR_LET@84..171
                    LetKeyword@84..87 "let"
                    Whitespace@87..88 " "
                    PATTERN_WILDCARD@88..90
                      WildcardKeyword@88..89 "_"
                      Whitespace@89..90 " "
                    Eq@90..91 "="
                    Whitespace@91..92 " "
                    EXPR_LET_VALUE@92..128
                      EXPR_CALL@92..128
                        EXPR_LIDENT@92..106
                          Lident@92..106 "string_println"
                        ARG_LIST@106..128
                          LParen@106..107 "("
                          ARG@107..126
                            EXPR_CALL@107..126
                              EXPR_LIDENT@107..120
                                Lident@107..120 "int_to_string"
                              ARG_LIST@120..126
                                LParen@120..121 "("
                                ARG@121..125
                                  EXPR_CALL@121..125
                                    EXPR_LIDENT@121..122
                                      Lident@121..122 "f"
                                    ARG_LIST@122..125
                                      LParen@122..123 "("
                                      ARG@123..124
                                        EXPR_INT@123..124
                                          Int@123..124 "2"
                                      RParen@124..125 ")"
                                RParen@125..126 ")"
                          RParen@126..127 ")"
                          Whitespace@127..128 " "
                    InKeyword@128..130 "in"
                    Whitespace@130..135 "\n    "
                    EXPR_LET_BODY@135..171
                      EXPR_CALL@135..171
                        EXPR_LIDENT@135..149
                          Lident@135..149 "string_println"
                        ARG_LIST@149..171
                          LParen@149..150 "("
                          ARG@150..169
                            EXPR_CALL@150..169
                              EXPR_LIDENT@150..163
                                Lident@150..163 "int_to_string"
                              ARG_LIST@163..169
                                LParen@163..164 "("
                                ARG@164..168
                                  EXPR_CALL@164..168
                                    EXPR_LIDENT@164..165
                                      Lident@164..165 "f"
                                    ARG_LIST@165..168
                                      LParen@165..166 "("
                                      ARG@166..167
                                        EXPR_INT@166..167
                                          Int@166..167 "3"
                                      RParen@167..168 ")"
                                RParen@168..169 ")"
                          RParen@169..170 ")"
                          Whitespace@170..171 "\n"
      RBrace@171..172 "}"
      Whitespace@172..174 "\n\n"
  FN@174..236
    FnKeyword@174..176 "fn"
    Whitespace@176..177 " "
    Lident@177..188 "call_int_id"
    PARAM_LIST@188..215
      LParen@188..189 "("
      PARAM@189..207
        Lident@189..190 "f"
        Whitespace@190..191 " "
        Colon@191..192 ":"
        Whitespace@192..193 " "
        TYPE_FUNC@193..205
          TYPE_TUPLE@193..199
            TYPE_LIST@193..199
              LParen@193..194 "("
              TYPE_INT@194..197
                IntKeyword@194..197 "int"
              RParen@197..198 ")"
              Whitespace@198..199 " "
          Arrow@199..201 "->"
          Whitespace@201..202 " "
          TYPE_INT@202..205
            IntKeyword@202..205 "int"
        Comma@205..206 ","
        Whitespace@206..207 " "
      PARAM@207..213
        Lident@207..208 "v"
        Colon@208..209 ":"
        Whitespace@209..210 " "
        TYPE_INT@210..213
          IntKeyword@210..213 "int"
      RParen@213..214 ")"
      Whitespace@214..215 " "
    Arrow@215..217 "->"
    Whitespace@217..218 " "
    TYPE_INT@218..222
      IntKeyword@218..221 "int"
      Whitespace@221..222 " "
    BLOCK@222..236
      LBrace@222..223 "{"
      Whitespace@223..228 "\n    "
      EXPR_CALL@228..233
        EXPR_LIDENT@228..229
          Lident@228..229 "f"
        ARG_LIST@229..233
          LParen@229..230 "("
          ARG@230..231
            EXPR_LIDENT@230..231
              Lident@230..231 "v"
          RParen@231..232 ")"
          Whitespace@232..233 "\n"
      RBrace@233..234 "}"
      Whitespace@234..236 "\n\n"
  FN@236..738
    FnKeyword@236..238 "fn"
    Whitespace@238..239 " "
    Lident@239..243 "main"
    PARAM_LIST@243..246
      LParen@243..244 "("
      RParen@244..245 ")"
      Whitespace@245..246 " "
    BLOCK@246..738
      LBrace@246..247 "{"
      Whitespace@247..252 "\n    "
      EXPR_LET@252..736
        LetKeyword@252..255 "let"
        Whitespace@255..256 " "
        PATTERN_VARIABLE@256..261
          Lident@256..260 "base"
          Whitespace@260..261 " "
        Eq@261..262 "="
        Whitespace@262..263 " "
        EXPR_LET_VALUE@263..265
          EXPR_INT@263..265
            Int@263..264 "5"
            Whitespace@264..265 " "
        InKeyword@265..267 "in"
        Whitespace@267..272 "\n    "
        EXPR_LET_BODY@272..736
          EXPR_LET@272..736
            LetKeyword@272..275 "let"
            Whitespace@275..276 " "
            PATTERN_VARIABLE@276..285
              Lident@276..284 "add_base"
              Whitespace@284..285 " "
            Eq@285..286 "="
            Whitespace@286..287 " "
            EXPR_LET_VALUE@287..305
              EXPR_CLOSURE@287..305
                CLOSURE_PARAM_LIST@287..296
                  Pipe@287..288 "|"
                  CLOSURE_PARAM@288..294
                    PATTERN_VARIABLE@288..289
                      Lident@288..289 "x"
                    Colon@289..290 ":"
                    Whitespace@290..291 " "
                    TYPE_INT@291..294
                      IntKeyword@291..294 "int"
                  Pipe@294..295 "|"
                  Whitespace@295..296 " "
                EXPR_CLOSURE_BODY@296..305
                  EXPR_BINARY@296..305
                    EXPR_LIDENT@296..298
                      Lident@296..297 "x"
                      Whitespace@297..298 " "
                    Plus@298..299 "+"
                    Whitespace@299..300 " "
                    EXPR_LIDENT@300..305
                      Lident@300..304 "base"
                      Whitespace@304..305 " "
            InKeyword@305..307 "in"
            Whitespace@307..312 "\n    "
            EXPR_LET_BODY@312..736
              EXPR_LET@312..736
                LetKeyword@312..315 "let"
                Whitespace@315..316 " "
                PATTERN_VARIABLE@316..323
                  Lident@316..322 "result"
                  Whitespace@322..323 " "
                Eq@323..324 "="
                Whitespace@324..325 " "
                EXPR_LET_VALUE@325..337
                  EXPR_CALL@325..337
                    EXPR_LIDENT@325..333
                      Lident@325..333 "add_base"
                    ARG_LIST@333..337
                      LParen@333..334 "("
                      ARG@334..335
                        EXPR_INT@334..335
                          Int@334..335 "7"
                      RParen@335..336 ")"
                      Whitespace@336..337 " "
                InKeyword@337..339 "in"
                Whitespace@339..344 "\n    "
                EXPR_LET_BODY@344..736
                  EXPR_LET@344..736
                    LetKeyword@344..347 "let"
                    Whitespace@347..348 " "
                    PATTERN_VARIABLE@348..356
                      Lident@348..355 "printer"
                      Whitespace@355..356 " "
                    Eq@356..357 "="
                    Whitespace@357..358 " "
                    EXPR_LET_VALUE@358..493
                      EXPR_CLOSURE@358..493
                        CLOSURE_PARAM_LIST@358..387
                          Pipe@358..359 "|"
                          CLOSURE_PARAM@359..373
                            PATTERN_VARIABLE@359..365
                              Lident@359..365 "prefix"
                            Colon@365..366 ":"
                            Whitespace@366..367 " "
                            TYPE_STRING@367..373
                              StringKeyword@367..373 "string"
                          Comma@373..374 ","
                          Whitespace@374..375 " "
                          CLOSURE_PARAM@375..385
                            PATTERN_VARIABLE@375..380
                              Lident@375..380 "value"
                            Colon@380..381 ":"
                            Whitespace@381..382 " "
                            TYPE_INT@382..385
                              IntKeyword@382..385 "int"
                          Pipe@385..386 "|"
                          Whitespace@386..387 " "
                        EXPR_CLOSURE_BODY@387..493
                          BLOCK@387..493
                            LBrace@387..388 "{"
                            Whitespace@388..397 "\n        "
                            EXPR_LET@397..491
                              LetKeyword@397..400 "let"
                              Whitespace@400..401 " "
                              PATTERN_VARIABLE@401..409
                                Lident@401..408 "message"
                                Whitespace@408..409 " "
                              Eq@409..410 "="
                              Whitespace@410..411 " "
                              EXPR_LET_VALUE@411..452
                                EXPR_CALL@411..452
                                  EXPR_LIDENT@411..421
                                    Lident@411..421 "string_add"
                                  ARG_LIST@421..452
                                    LParen@421..422 "("
                                    ARG@422..430
                                      EXPR_LIDENT@422..428
                                        Lident@422..428 "prefix"
                                      Comma@428..429 ","
                                      Whitespace@429..430 " "
                                    ARG@430..450
                                      EXPR_CALL@430..450
                                        EXPR_LIDENT@430..443
                                          Lident@430..443 "int_to_string"
                                        ARG_LIST@443..450
                                          LParen@443..444 "("
                                          ARG@444..449
                                            EXPR_LIDENT@444..449
                                              Lident@444..449 "value"
                                          RParen@449..450 ")"
                                    RParen@450..451 ")"
                                    Whitespace@451..452 " "
                              InKeyword@452..454 "in"
                              Whitespace@454..463 "\n        "
                              EXPR_LET_BODY@463..491
                                EXPR_CALL@463..491
                                  EXPR_LIDENT@463..477
                                    Lident@463..477 "string_println"
                                  ARG_LIST@477..491
                                    LParen@477..478 "("
                                    ARG@478..485
                                      EXPR_LIDENT@478..485
                                        Lident@478..485 "message"
                                    RParen@485..486 ")"
                                    Whitespace@486..491 "\n    "
                            RBrace@491..492 "}"
                            Whitespace@492..493 " "
                    InKeyword@493..495 "in"
                    Whitespace@495..500 "\n    "
                    EXPR_LET_BODY@500..736
                      EXPR_LET@500..736
                        LetKeyword@500..503 "let"
                        Whitespace@503..504 " "
                        PATTERN_WILDCARD@504..506
                          WildcardKeyword@504..505 "_"
                          Whitespace@505..506 " "
                        Eq@506..507 "="
                        Whitespace@507..508 " "
                        EXPR_LET_VALUE@508..536
                          EXPR_CALL@508..536
                            EXPR_LIDENT@508..515
                              Lident@508..515 "printer"
                            ARG_LIST@515..536
                              LParen@515..516 "("
                              ARG@516..528
                                EXPR_STR@516..526
                                  Str@516..526 "\"result: \""
                                Comma@526..527 ","
                                Whitespace@527..528 " "
                              ARG@528..534
                                EXPR_LIDENT@528..534
                                  Lident@528..534 "result"
                              RParen@534..535 ")"
                              Whitespace@535..536 " "
                        InKeyword@536..538 "in"
                        Whitespace@538..543 "\n    "
                        EXPR_LET_BODY@543..736
                          EXPR_LET@543..736
                            LetKeyword@543..546 "let"
                            Whitespace@546..547 " "
                            PATTERN_VARIABLE@547..554
                              Lident@547..553 "unused"
                              Whitespace@553..554 " "
                            Eq@554..555 "="
                            Whitespace@555..556 " "
                            EXPR_LET_VALUE@556..576
                              EXPR_CLOSURE@556..576
                                CLOSURE_PARAM_LIST@556..565
                                  Pipe@556..557 "|"
                                  CLOSURE_PARAM@557..563
                                    PATTERN_VARIABLE@557..558
                                      Lident@557..558 "y"
                                    Colon@558..559 ":"
                                    Whitespace@559..560 " "
                                    TYPE_INT@560..563
                                      IntKeyword@560..563 "int"
                                  Pipe@563..564 "|"
                                  Whitespace@564..565 " "
                                EXPR_CLOSURE_BODY@565..576
                                  EXPR_BINARY@565..576
                                    EXPR_LIDENT@565..567
                                      Lident@565..566 "y"
                                      Whitespace@566..567 " "
                                    Plus@567..568 "+"
                                    Whitespace@568..569 " "
                                    EXPR_LIDENT@569..576
                                      Lident@569..575 "result"
                                      Whitespace@575..576 " "
                            InKeyword@576..578 "in"
                            Whitespace@578..583 "\n    "
                            EXPR_LET_BODY@583..736
                              EXPR_LET@583..736
                                LetKeyword@583..586 "let"
                                Whitespace@586..587 " "
                                PATTERN_VARIABLE@587..598
                                  Lident@587..597 "no_capture"
                                  Whitespace@597..598 " "
                                Eq@598..599 "="
                                Whitespace@599..600 " "
                                EXPR_LET_VALUE@600..615
                                  EXPR_CLOSURE@600..615
                                    CLOSURE_PARAM_LIST@600..609
                                      Pipe@600..601 "|"
                                      CLOSURE_PARAM@601..607
                                        PATTERN_VARIABLE@601..602
                                          Lident@601..602 "z"
                                        Colon@602..603 ":"
                                        Whitespace@603..604 " "
                                        TYPE_INT@604..607
                                          IntKeyword@604..607 "int"
                                      Pipe@607..608 "|"
                                      Whitespace@608..609 " "
                                    EXPR_CLOSURE_BODY@609..615
                                      EXPR_BINARY@609..615
                                        EXPR_LIDENT@609..611
                                          Lident@609..610 "z"
                                          Whitespace@610..611 " "
                                        Star@611..612 "*"
                                        Whitespace@612..613 " "
                                        EXPR_INT@613..615
                                          Int@613..614 "2"
                                          Whitespace@614..615 " "
                                InKeyword@615..617 "in"
                                Whitespace@617..622 "\n    "
                                EXPR_LET_BODY@622..736
                                  EXPR_LET@622..736
                                    LetKeyword@622..625 "let"
                                    Whitespace@625..626 " "
                                    PATTERN_VARIABLE@626..634
                                      Lident@626..633 "doubled"
                                      Whitespace@633..634 " "
                                    Eq@634..635 "="
                                    Whitespace@635..636 " "
                                    EXPR_LET_VALUE@636..650
                                      EXPR_CALL@636..650
                                        EXPR_LIDENT@636..646
                                          Lident@636..646 "no_capture"
                                        ARG_LIST@646..650
                                          LParen@646..647 "("
                                          ARG@647..648
                                            EXPR_INT@647..648
                                              Int@647..648 "3"
                                          RParen@648..649 ")"
                                          Whitespace@649..650 " "
                                    InKeyword@650..652 "in"
                                    Whitespace@652..657 "\n    "
                                    EXPR_LET_BODY@657..736
                                      EXPR_LET@657..736
                                        LetKeyword@657..660 "let"
                                        Whitespace@660..661 " "
                                        PATTERN_WILDCARD@661..663
                                          WildcardKeyword@661..662 "_"
                                          Whitespace@662..663 " "
                                        Eq@663..664 "="
                                        Whitespace@664..665 " "
                                        EXPR_LET_VALUE@665..704
                                          EXPR_CALL@665..704
                                            EXPR_LIDENT@665..679
                                              Lident@665..679 "string_println"
                                            ARG_LIST@679..704
                                              LParen@679..680 "("
                                              ARG@680..702
                                                EXPR_CALL@680..702
                                                  EXPR_LIDENT@680..693
                                                    Lident@680..693 "int_to_string"
                                                  ARG_LIST@693..702
                                                    LParen@693..694 "("
                                                    ARG@694..701
                                                      EXPR_LIDENT@694..701
                                                        Lident@694..701 "doubled"
                                                    RParen@701..702 ")"
                                              RParen@702..703 ")"
                                              Whitespace@703..704 " "
                                        InKeyword@704..706 "in"
                                        Whitespace@706..711 "\n    "
                                        EXPR_LET_BODY@711..736
                                          EXPR_LET@711..736
                                            LetKeyword@711..714 "let"
                                            Whitespace@714..715 " "
                                            PATTERN_WILDCARD@715..717
                                              WildcardKeyword@715..716 "_"
                                              Whitespace@716..717 " "
                                            Eq@717..718 "="
                                            Whitespace@718..719 " "
                                            EXPR_LET_VALUE@719..726
                                              EXPR_CALL@719..726
                                                EXPR_LIDENT@719..723
                                                  Lident@719..723 "test"
                                                ARG_LIST@723..726
                                                  LParen@723..724 "("
                                                  RParen@724..725 ")"
                                                  Whitespace@725..726 " "
                                            InKeyword@726..728 "in"
                                            Whitespace@728..733 "\n    "
                                            EXPR_LET_BODY@733..736
                                              EXPR_UNIT@733..736
                                                LParen@733..734 "("
                                                RParen@734..735 ")"
                                                Whitespace@735..736 "\n"
      RBrace@736..737 "}"
      Whitespace@737..738 "\n"