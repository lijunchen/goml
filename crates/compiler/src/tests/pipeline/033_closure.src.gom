package main

import (
    "fmt"
)

func int_to_string(x int) string {
    return fmt.Sprintf("%d", x)
}

func int_add(x int, y int) int {
    return x + y
}

func int_mul(x int, y int) int {
    return x * y
}

func string_add(a string, b string) string {
    return a + b
}

func string_println(s string) struct{} {
    fmt.Println(s)
    return struct{}{}
}

type Point struct {
    x int
    y int
}

type closure_env_f_0 struct {
    y_0 int
    z_1 int
}

type closure_env_add_base_1 struct {
    base_0 int
}

type closure_env_printer_2 struct {}

type closure_env_unused_3 struct {
    result_0 int
}

type closure_env_no_capture_4 struct {}

type closure_env_play_list_and_point_5 struct {
    list123_0 IntList
    point_1 Point
}

type IntList interface {
    isIntList()
}

type Nil struct {}

func (_ Nil) isIntList() {}

type Cons struct {
    _0 int
    _1 IntList
}

func (_ Cons) isIntList() {}

func test() struct{} {
    var ret40 struct{}
    var y__0 int = 3
    var z__1 int = 5
    var f__3 closure_env_f_0 = closure_env_f_0{
        y_0: y__0,
        z_1: z__1,
    }
    var t24 int = __closure_apply11(f__3, 2)
    var t23 string = int_to_string(t24)
    string_println(t23)
    var t26 int = __closure_apply11(f__3, 3)
    var t25 string = int_to_string(t26)
    ret40 = string_println(t25)
    return ret40
}

func main0() struct{} {
    var ret42 struct{}
    var base__6 int = 5
    var add_base__8 closure_env_add_base_1 = closure_env_add_base_1{
        base_0: base__6,
    }
    var result__9 int = __closure_apply13(add_base__8, 7)
    var printer__13 closure_env_printer_2 = closure_env_printer_2{}
    __closure_apply15(printer__13, "result: ", result__9)
    var no_capture__17 closure_env_no_capture_4 = closure_env_no_capture_4{}
    var doubled__18 int = __closure_apply19(no_capture__17, 3)
    var t27 string = int_to_string(doubled__18)
    string_println(t27)
    test()
    var t30 IntList = Nil{}
    var t29 IntList = Cons{
        _0: 3,
        _1: t30,
    }
    var t28 IntList = Cons{
        _0: 2,
        _1: t29,
    }
    var list123__19 IntList = Cons{
        _0: 1,
        _1: t28,
    }
    var point__20 Point = Point{
        x: 10,
        y: 20,
    }
    var play_list_and_point__25 closure_env_play_list_and_point_5 = closure_env_play_list_and_point_5{
        list123_0: list123__19,
        point_1: point__20,
    }
    __closure_apply21(play_list_and_point__25)
    ret42 = struct{}{}
    return ret42
}

func __closure_apply11(env12 closure_env_f_0, x__2 int) int {
    var ret43 int
    var y__0 int = env12.y_0
    var z__1 int = env12.z_1
    var t31 int = int_mul(x__2, y__0)
    ret43 = int_mul(t31, z__1)
    return ret43
}

func __closure_apply13(env14 closure_env_add_base_1, x__7 int) int {
    var ret44 int
    var base__6 int = env14.base_0
    ret44 = int_add(x__7, base__6)
    return ret44
}

func __closure_apply15(env16 closure_env_printer_2, prefix__10 string, value__11 int) struct{} {
    var ret45 struct{}
    var t32 string = int_to_string(value__11)
    var message__12 string = string_add(prefix__10, t32)
    ret45 = string_println(message__12)
    return ret45
}

func __closure_apply19(env20 closure_env_no_capture_4, z__16 int) int {
    var ret47 int
    ret47 = int_mul(z__16, 2)
    return ret47
}

func __closure_apply21(env22 closure_env_play_list_and_point_5) struct{} {
    var ret48 struct{}
    var list123__19 IntList = env22.list123_0
    var point__20 Point = env22.point_1
    switch list123__19 := list123__19.(type) {
    case Nil:
        ret48 = string_println("Empty list")
    case Cons:
        var x4 int = list123__19._0
        var head__21 int = x4
        var t33 string = int_to_string(head__21)
        string_println(t33)
        var x7 int = point__20.x
        var x8 int = point__20.y
        var y__24 int = x8
        var x__23 int = x7
        var t36 string = int_to_string(x__23)
        var t39 string = int_to_string(y__24)
        var t38 string = string_add(t39, ")")
        var t37 string = string_add(", ", t38)
        var t35 string = string_add(t36, t37)
        var t34 string = string_add("Point: (", t35)
        string_println(t34)
        ret48 = struct{}{}
    }
    return ret48
}

func main() {
    main0()
}
