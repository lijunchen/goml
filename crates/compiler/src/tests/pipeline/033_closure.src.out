result: 12
6
30
45
1
Point: (10, 20)
